WPThemeReview/Sniffs/Privacy/ShortenedURLsSniff.php

```diff
@@ -0,0 +1,122 @@
+<?php
+/**
+ * WPThemeReview Coding Standard.
+ *
+ * @package WPTRT\WPThemeReview
+ * @link    https://github.com/WPTRT/WPThemeReview
+ * @license https://opensource.org/licenses/MIT MIT
+ */
+
+namespace WPThemeReview\Sniffs\Privacy;
+
+use PHP_CodeSniffer\Sniffs\Sniff;
+use PHP_CodeSniffer\Files\File;
+use PHP_CodeSniffer\Util\Tokens;
+
+/**
+ * Check if the file contains a shortened URL from the list of banned URL shortener services.
+ *
+ * @since 0.2.0
+ */
+class ShortenedURLsSniff implements Sniff {
+
+	/**
+	 * Error message template.
+	 *
+	 * @since 0.2.0
+	 *
+	 * @var string
+	 */
+	const ERROR_MSG = 'Shortened URLs are not allowed in the theme. Found: "%s".';
+
+	/**
+	 * Found used shortener in a file
+	 *
+	 * @since 0.2.0
+	 *
+	 * @var string
+	 */
+	protected $shortener;
+
+	/**
+	 * Supported Tokenizers
+	 *
+	 * @since 0.2.0
+	 *
+	 * @var array
+	 */
+	public $supportedTokenizers = array(
+		'PHP',
+		'CSS',
+		'JS',
+	);
+
+	/**
+	 * List of url shorteners.
+	 *
+	 * @since 0.2.0
+	 *
+	 * @var array
+	 */
+	protected $url_shorteners = [
+		'bit.do',
+		'bit.ly',
+		'df.ly',
+		'goo.gl',
+		'is.gd',
+		'lc.chat',
+		'ow.ly',
+		'polr.me',
+		's2r.co',
+		'soo.gd',
+		'tiny.cc',
+		'tinyurl.com',
+	];
+
+	/**
+	 * Returns an array of tokens this test wants to listen for.
+	 *
+	 * @since 0.2.0
+	 *
+	 * @return array
+	 */
+	public function register() {
+		return Tokens::$textStringTokens + array(
+			T_COMMENT,
+			T_DOC_COMMENT_STRING,
+			T_DOC_COMMENT,
+		);
+	}
+
+	/**
+	 * Processes this test, when one of its tokens is encountered.
+	 *
+	 * @since 0.2.0
+	 *
+	 * @param \PHP_CodeSniffer\Files\File $phpcsFile The PHP_CodeSniffer file where the
+	 *                                               token was found.
+	 * @param int                         $stackPtr  The position of the current token
+	 *                                               in the stack.
+	 *
+	 * @return void
+	 */
+	public function process( File $phpcsFile, $stackPtr ) {
+		$tokens  = $phpcsFile->getTokens();
+		$content = $tokens[ $stackPtr ]['content'];
+
+		if ( strpos( $content, '.' ) === false ) {
+			return;
+		}
+
+		foreach ( $this->url_shorteners as $url_shortener ) {
+			if ( strpos( $content, $url_shortener ) !== false ) {
```

Comment:
Yeah, I can see your point. Mentioning them is ok, using is not. I'll give the regex check a go :+1:

Reply:
Note: I've edited the above code sample and added the `i` modifier to the regex template.

Reply:
Yup, noticed that when I ran the tests and it wasn't picking up the case insensitive links. Also the single quote inside the `[^\s'"]` shouldn't be there, right? I've also added 

```php
$this->regex = sprintf(
    self::REGEX_TEMPLATE,
    implode( '/|', $urls )
);
```

becuase without it, the `esc_html_e( 'You can use bit.ly or goo.gl to generate short URLs', 'text-domain' );` was picked up as an error, and it's not (missing `/`).

Reply:
I've added 

```php

echo <<<'EOD'
Example of string spanning multiple lines
iS.gd/O9iJkpsj is some string.
EOD; // Bad.

echo <<<EOT
My name is "$name". I am printing some soo.gd/pOlt6gjka.
This should print a capital 'A': \x41
EOT;
} // Bad.
```

To the tests but they are not picked up. And when I added inline html after those, the sniffer didn't pick it up.

Is this a bug in the phpcs or am I doing something wrong? @jrfnl

Reply:
>  Also the single quote inside the `[^\s'"]` shouldn't be there, right?

IIRC a single quote is not allowed to be in a URL, nor the double quote, which is why I listed them here, but checking this with the RFC for URLs can't hurt.

If the single quote remains, then it should be escaped `[^\s\'"]` as the constant is a single quoted string (I did mention the above code sample was pseudo code/untested, didn't I ?)

> I've also added
> ```php
> $this->regex = sprintf(
>     self::REGEX_TEMPLATE,
>     implode( '/|', $urls )
> );
> ```
> becuase without it, the esc_html_e( 'You can use bit.ly or goo.gl to generate short URLs', 'text-domain' ); was picked up as an error, and it's not (missing /).

This is incorrect.
Again - the above code sample was pseudo code/untested.
The problem was not the `implode()`, but the regex template. When this is changed to `'`(?:%s)/[^\s\'"]+`i'`, it should be fine.

(Let me know if you'd like me to explain the change in the regex)

> Also for some reason, inline html, heredoc and nowdoc aren't being picked up (they should be since they are in Tokens::$textStringTokens). Any ideas what am I missing?

I couldn't reproduce that, but have taken the liberty to clean up the test file a little now anyway.
Note: when there are (unintentional) parse errors in the PHP test code file, the tokenizer may give incorrect results, especially so with Heredoc/Nowdoc.

Reply:
For the regex, the first part will create a non capturing group made up of the urls we want to mach, the second part matches anything after the `/` in the url (unlimited amount of times because of +), except the whitespace character, `'` and `"`, and the `i` is for the case insensitivity. 

I'm usually using https://regex101.com/ to help me understand the regexes 😄

Reply:
And the missing group parentheses was causing the problem you ran into earlier.